distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso

```diff
@@ -1,99 +1,307 @@
 import project.Any.Any
-import project.Data.Boolean.Boolean
-import project.Data.Json.JS_Object
-import project.Data.Numbers.Float
-import project.Data.Numbers.Integer
+import project.Data.Range.Range
 import project.Data.Text.Text
+import project.Data.Time.Date.Date
+import project.Data.Time.Date_Range.Date_Range
+import project.Data.Time.Time_Of_Day.Time_Of_Day
+import project.Data.Time.Time_Period.Time_Period
+import project.Data.Map.Map
+import project.Data.Set.Set
 import project.Data.Vector.Vector
 import project.Error.Error
 import project.Errors.Illegal_Argument.Illegal_Argument
+import project.Nothing.Nothing
+import project.Runtime.Ref.Ref
 import project.System
+from project.Data.Boolean import Boolean, False, True
+from project.Data.Numbers import Float, Integer
+from project.Data.Range.Extensions import all
 
 polyglot java import java.lang.Integer as Java_Integer
 polyglot java import java.util.Random as Java_Random
 polyglot java import java.util.UUID
 polyglot java import org.enso.base.Random_Utils
 
-## PRIVATE
-   UNSTABLE
-   Returns a default seed to use for random number generation.
+type Random
+    ## PRIVATE
+       GROUP Random
 
-   The returned seed may differ between each call to this method.
-get_default_seed : Integer
-get_default_seed = System.nano_time
+       Create a new `Random_Generator` with the specified seed.
+
+       Arguments:
+       - seed: The seed used to initialze the `Random_Generator`.
+
+       > Example
+         Create a new `Random_Generator` and use it.
+
+             import Standard.Base.Random.Random
+
+             rng = Random.new_generator 42
+             i = rng.integer 0 10
+    new_generator : Integer -> Random_Generator
+    new_generator (seed:Integer=get_default_seed) = Random_Generator.new seed
+
+    ## TEXT_ONLY
+       GROUP Random
+
+       Set the seed of the default `Random_Generator` instance.
+
+       Arguments:
+       - seed: The seed used to initialze the default `Random_Generator`.
+
+       > Example
+         Set the seed of the default `Random_Generator` to 42.
 
-## Constructs a new random number generator.
-new : Integer -> Random_Number_Generator
-new seed=get_default_seed =
-    Random_Number_Generator.Value (Java_Random.new seed)
+             import Standard.Base.Random.Random
 
-type Random_Number_Generator
-    ## A random number generator.
-    Value java_random
+             Random.set_seed 42
+    set_seed : Integer -> Nothing
+    set_seed (seed:Integer=get_default_seed) = Random_Utils.seedRandomsSingleton seed
```

Comment:
_"static setter"_!? That's an API design antipattern even in Java - not talking about functional language. I'd like to read justification why is it a good idea?

Reply:
I couldn't find a way to do this in pure Enso and @jdunkerley confirmed that it's not possible, and must be done in Java.

The reason this is needed: the design decision was that we have an implicit singleton generator called via static method, e.g. `Random.integer 0 10`. The static method delegates to the singleton instance.

Thus the static setter. Global mutable state is terrible, but in this narrow use it's exactly what we want. And the setter is only required because we can't define a global singleton in pure Enso.

Later, we will allow users to create generator instances; we are not allowing that yet because it is unclear how to use them properly in the IDE, since they are stateful and the Enso graph does not easily allow fine control of stateful objects. `Random_Generator` is the instance, but it is not meant to be visible yet.

I don't think the filename is a good place to encode that the class is hidden; we generally move private classes to an `Internal` subdirectory so I will do that.

Reply:
Reverted, since I need to access this from tests.

Reply:
>  we can't define a global singleton

That's by design, I assume.